include/boost/iostreams/filter/zlib.hpp

```diff
@@ -381,7 +381,8 @@ bool zlib_decompressor_impl<Alloc>::filter
     int result = xinflate(zlib::sync_flush);
     after(src_begin, dest_begin, false);
     zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(result);
-    return !(eof_ = result == zlib::stream_end);
+    eof_ = result == zlib::stream_end;
+    return result == zlib::okay;
```

Comment:
The return value is used to check the success of the operation, not to check the stream_end. Instead the eof_ variable is used to check the end of stream case.